tests/epics/adcore/test_drivers.py

```diff
@@ -93,3 +93,48 @@ async def wait_then_fail():
         ValueError, match="Final detector state Disconnected not in valid end states:"
     ):
         await acquiring
+
+
+@patch("ophyd_async.core._detector.DEFAULT_TIMEOUT", 0.2)
+async def test_start_acquiring_driver_and_ensure_status_timing(
+    controller: adsimdetector.SimController,
+):
+    """This test ensures the camera has time to return to a good state.
+
+    Real world application; there is race condition wherein the
+    detector has been asked to complete acquisition, but has not yet
+    returned to a known good state before the status check.
+
+    """
+    set_mock_value(controller.driver.detector_state, adcore.ADState.ACQUIRE)
+
+    acquiring = await controller.start_acquiring_driver_and_ensure_status()
+
+    async def complete_acquire():
+        """Return to idle state, but pretend the detector is slow."""
+        await asyncio.sleep(0.05)
+        set_mock_value(controller.driver.detector_state, adcore.ADState.IDLE)
+
+    await asyncio.gather(acquiring, complete_acquire())
+
+
+async def bad_observe_value(*args, **kwargs):
+    "Stub to simulate a disconnected ``observe_value()``."
+    if True:
+        raise asyncio.TimeoutError()
+    yield None  # Make it a generator
+
+
+@patch("ophyd_async.epics.adcore._core_logic.observe_value", bad_observe_value)
+@patch("ophyd_async.core._detector.DEFAULT_TIMEOUT", 0.2)
+async def test_start_acquiring_driver_and_ensure_status_disconnected(
+    controller: adsimdetector.SimController,
+):
+    """This test ensures the function behaves gracefully if no detector
+    states are available.
+
+    """
+    acquiring = await controller.start_acquiring_driver_and_ensure_status()
+
+    with pytest.raises(asyncio.TimeoutError):
```

Comment:
Please could we check the error message here

Reply:
Sure. The original ``TimeoutError`` exception does not have a message. I put in a message when re-raising the exception that provides the source of the detector state signal.

Is it safe to assume that `driver.detector_state` will always have a `.source`?

Reply:
Yes